### <font color='red'> Project 2

Project Description:
- Use same datasets as Project 1.
- Preprocess data: Explore data and apply data scaling.

Regression Task:
- Apply any two models with bagging and any two models with pasting.
- Apply any two models with adaboost boosting
- Apply one model with gradient boosting
- Apply PCA on data and then apply all the models in project 1 again on data you get from PCA. Compare your results with results in project 2. You don't need to apply all the models twice. Just copy the result table from project 1, prepare similar table for all the models after PCA and compare both tables. Does PCA help in getting better results?
- Apply deep learning models covered in class

Classification Task:
- Apply two voting classifiers - one with hard voting and one with soft voting
- Apply any two models with bagging and any two models with pasting.
- Apply any two models with adaboost boosting
- Apply one model with gradient boosting
- Apply PCA on data and then apply all the models in project 1 again on data you get from PCA. Compare your results with results in project 1. You don't need to apply all the models twice. Just copy the result table from project 1, prepare similar table for all the models after PCA and compare both tables. Does PCA help in getting better results?
- Apply deep learning models covered in class

Deliverables:
- Use markdown to provide inline comments for this project.
- Your outputs should be clearly executed in the notebook i.e. we should not need to rerun the code to obtain the outputs.
- Visualization encouraged.
- If you are submitting two different files, then please only one group member submit both the files. If you submit two files separately from different accounts, it will be submitted as two different attempts.
- If you are submitting two different files, then please follow below naming convetion:
    Project2_Regression_GroupXX_Firstname1_Firstname2.ipynb
    Project2_Classification_GroupXX_Firstname1_Firstname2.ipynb
- If you are submitting single file, then please follow below naming convetion:
    Project2_Both_GroupXX_Firstname1_Firstname2.ipynb

Questions regarding the project:
- We have created a discussion board under Projects folder on e-learning. Create threads over there and post your queries related to project there.
- We will also answer queries there. We will not be answering any project related queries through the mail.

In [78]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [79]:
audit_risk = pd.read_csv('audit_risk.csv')
trial = pd.read_csv('trial.csv')

In [80]:
df = pd.merge(audit_risk, trial, how='left')

In [81]:
df.dtypes

Sector_score      float64
LOCATION_ID        object
PARA_A            float64
Score_A           float64
Risk_A            float64
PARA_B            float64
Score_B           float64
Risk_B            float64
TOTAL             float64
numbers           float64
Score_B.1         float64
Risk_C            float64
Money_Value       float64
Score_MV          float64
Risk_D            float64
District_Loss       int64
PROB              float64
RiSk_E            float64
History             int64
Prob              float64
Risk_F            float64
Score             float64
Inherent_Risk     float64
CONTROL_RISK      float64
Detection_Risk    float64
Audit_Risk        float64
Risk                int64
SCORE_A           float64
SCORE_B           float64
Marks             float64
MONEY_Marks       float64
District          float64
Loss              float64
LOSS_SCORE        float64
History_score     float64
dtype: object

In [82]:
# Delete the records with non-numerical values and convert the LOCATION_ID into numerical data
df = df[df.LOCATION_ID.apply(lambda x: x.isnumeric())]
df.LOCATION_ID = df.LOCATION_ID.apply(eval)

In [83]:
df.head(5)

Sector_score  LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  \
0          3.89           23    4.18      0.6   2.508    2.50      0.2   
1          3.89            6    0.00      0.2   0.000    4.83      0.2   
2          3.89            6    0.51      0.2   0.102    0.23      0.2   
3          3.89            6    0.00      0.2   0.000   10.80      0.6   
4          3.89            6    0.00      0.2   0.000    0.08      0.2   

   Risk_B  TOTAL  numbers      ...        Audit_Risk  Risk  SCORE_A  SCORE_B  \
0   0.500   6.68      5.0      ...            1.7148     1      6.0      2.0   
1   0.966   4.83      5.0      ...            0.5108     0      2.0      2.0   
2   0.046   0.74      5.0      ...            0.3096     0      2.0      2.0   
3   6.480  10.80      6.0      ...            3.5060     1      2.0      6.0   
4   0.016   0.08      5.0      ...            0.2832     0      2.0      2.0   

   Marks  MONEY_Marks  District  Loss  LOSS_SCORE  History_score  
0    2.0          2.0       2.0   0.0         2.0            2.0  
1    2.0          2.0       2.0   0.0         2.0            2.0  
2    2.0          2.0       2.0   0.0         2.0            2.0  
3    6.0          6.0       2.0   0.0         2.0            2.0  
4    2.0          2.0       2.0   0.0         2.0            2.0  

[5 rows x 35 columns]

In [84]:
# Score_A and Score_B contain the same information with SCORE_A and SCORE_B while the two pairs are on a different scale. So we
# keep SCORE_A and SCORE_B and delete Score_A and Score_B, in this way, we avoid data redundancy and also keep the two columns 
# in the same scale.
df.drop(['Score_A', 'Score_B'], axis=1, inplace = True)
df.dropna(axis = 0, inplace = True) # Here we drop the data with missing values.

In [85]:
df.shape

(625, 33)

In [86]:
df.dtypes

Sector_score      float64
LOCATION_ID         int64
PARA_A            float64
Risk_A            float64
PARA_B            float64
Risk_B            float64
TOTAL             float64
numbers           float64
Score_B.1         float64
Risk_C            float64
Money_Value       float64
Score_MV          float64
Risk_D            float64
District_Loss       int64
PROB              float64
RiSk_E            float64
History             int64
Prob              float64
Risk_F            float64
Score             float64
Inherent_Risk     float64
CONTROL_RISK      float64
Detection_Risk    float64
Audit_Risk        float64
Risk                int64
SCORE_A           float64
SCORE_B           float64
Marks             float64
MONEY_Marks       float64
District          float64
Loss              float64
LOSS_SCORE        float64
History_score     float64
dtype: object

# Regressions

In [87]:
X = df.drop(['Audit_Risk', 'Risk'], axis = 1)
y = df['Audit_Risk']

In [88]:
from sklearn.model_selection import train_test_split
X_train_org, X_test_org, y_train, y_test = train_test_split(X,y, random_state = 0)
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train_org)
X_test = scaler.transform(X_test_org)

## Bagging and Pasting on Linear Regression and Polynomial Regression

In [89]:
# Bagging on Linear Regression
from sklearn.ensemble import BaggingRegressor
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import GridSearchCV

param_grid = {'n_estimators': [100, 200, 300, 400, 500]}
lreg = LinearRegression()
bag_lreg = BaggingRegressor(lreg, bootstrap=True, random_state=0)
grid_search = GridSearchCV(bag_lreg, param_grid, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))
    

Best parameters: {'n_estimators': 200}
Best cross-validation score: -4401051843858142.50


In [90]:
bag_lreg1 = BaggingRegressor(lreg, n_estimators=200, bootstrap=True, random_state=0)
bag_lreg1.fit(X_train, y_train)
y_test_pred = bag_lreg1.predict(X_test)
from sklearn.metrics import r2_score
print("R_squred score for bagging linear regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for bagging linear regression with best parameters: 0.6044707375465956


In [91]:
# Bagging on Polynomial Regression

from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(2) # From project1 we know the best parameter for polynomial regression is n=2
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

param_grid = {'n_estimators': [100, 200, 300, 400, 500]}
lreg = LinearRegression()
bag_lreg = BaggingRegressor(lreg, bootstrap=True, random_state=0)
grid_search = GridSearchCV(bag_lreg, param_grid, cv=5, return_train_score=True)

grid_search.fit(X_train_poly, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'n_estimators': 100}
Best cross-validation score: 0.97


In [92]:
bag_poly1 = BaggingRegressor(lreg, n_estimators=100, bootstrap=True, random_state=0)
bag_poly1.fit(X_train_poly, y_train)
y_test_pred = bag_poly1.predict(X_test_poly)
from sklearn.metrics import r2_score
print("R_squred score for bagging polynomial regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for bagging polynomial regression with best parameters: 0.9978896899820212


In [93]:
# Pasting on Linear Regression

param_grid = {'n_estimators': [100, 200, 300, 400, 500]}
lreg = LinearRegression()
past_lreg = BaggingRegressor(lreg, bootstrap=False, random_state=0)
grid_search = GridSearchCV(past_lreg, param_grid, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))
    

Best parameters: {'n_estimators': 400}
Best cross-validation score: -72899493833805.97


In [94]:
past_lreg1 = BaggingRegressor(lreg, n_estimators=400, bootstrap=False, random_state=0)
past_lreg1.fit(X_train, y_train)
y_test_pred = past_lreg1.predict(X_test)
from sklearn.metrics import r2_score
print("R_squred score for pasting linear regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for pasting linear regression with best parameters: 0.4437886917390246


In [95]:
# Pasting on Polynomial Regression

from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(2) # From project1 we know the best parameter for polynomial regression is n=2
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

param_grid = {'n_estimators': [100, 200, 300, 400, 500]}
lreg = LinearRegression()
bag_lreg = BaggingRegressor(lreg, bootstrap=False, random_state=0)
grid_search = GridSearchCV(bag_lreg, param_grid, cv=5, return_train_score=True)
grid_search.fit(X_train_poly, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'n_estimators': 200}
Best cross-validation score: 0.91


In [96]:
bag_poly1 = BaggingRegressor(lreg, n_estimators=200, bootstrap=False, random_state=0)
bag_poly1.fit(X_train_poly, y_train)
y_test_pred = bag_poly1.predict(X_test_poly)

print("R_squred score for pasting polynomial regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for pasting polynomial regression with best parameters: 0.9992841739114662


## Adaboost boosting on Linear Regression and Polynomial Regression

In [97]:
# Adaboost on linear regression
from sklearn.ensemble import AdaBoostRegressor
lreg = LinearRegression()
param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
ada_reg = AdaBoostRegressor(lreg, random_state=0)
grid_search = GridSearchCV(ada_reg, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.2, 'n_estimators': 100}
Best cross-validation score: 0.24


In [98]:
ada_reg1 = AdaBoostRegressor(lreg, n_estimators=100, learning_rate=0.2, random_state=0)
ada_reg1.fit(X_train, y_train)
y_test_pred = ada_reg1.predict(X_test)
print("R_squred score for adaboost linear regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for adaboost linear regression with best parameters: 0.31808843737881387


In [99]:
# Adaboost on polynomial regression
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(2) # From project1 we know the best parameter for polynomial regression is n=2
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
ada_reg = AdaBoostRegressor(lreg, random_state=0)
grid_search = GridSearchCV(ada_reg, param, cv=5, return_train_score=True)
grid_search.fit(X_train_poly, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.4, 'n_estimators': 100}
Best cross-validation score: 0.97


In [100]:
ada_reg1 = AdaBoostRegressor(lreg, n_estimators=100, learning_rate=0.4, random_state=0)
ada_reg1.fit(X_train_poly, y_train)
y_test_pred = ada_reg1.predict(X_test_poly)
print("R_squred score for adaboost polynomial regression with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for adaboost polynomial regression with best parameters: 0.9990410846224799


## Gradient Boosting Model(Decision Tree Regressor)

In [101]:
from  sklearn.ensemble import GradientBoostingRegressor
param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
gbrt = GradientBoostingRegressor(random_state=0)
grid_search = GridSearchCV(gbrt, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.2, 'n_estimators': 500}
Best cross-validation score: 0.76


In [102]:
gbrt_reg1 = GradientBoostingRegressor(n_estimators=500, learning_rate=0.2, random_state=0)
gbrt_reg1.fit(X_train, y_train)
y_test_pred = gbrt_reg1.predict(X_test)
print("R_squred score for gradient boosting model with best parameters: {}".format(r2_score(y_test, y_test_pred)))

R_squred score for gradient boosting model with best parameters: 0.8980581256067349


## PCA and Regressions with best parameters

In [103]:
# Using grid search to find the best parameter for PCA
from sklearn.decomposition import PCA
param = {"n_components": list(range(10, 21))}
pca = PCA()
grid_search = GridSearchCV(pca, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'n_components': 10}
Best cross-validation score: 35.95


In [104]:
pca1 = PCA(n_components=10)
pca1.fit(X_train, y_train)
pca1.explained_variance_ratio_.sum()

0.9879864309421829

In [105]:
X_train_reduced = pca1.fit_transform(X_train)
X_test_reduced = pca1.transform(X_test)

In [106]:
# KNN regression after PCA

from sklearn.neighbors import KNeighborsRegressor
knn_reg3 = KNeighborsRegressor(3)
knn_reg3.fit(X_train_reduced, y_train)
y_pred = knn_reg3.predict(X_test_reduced)
print("R_squred score for KNN Regressor after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for KNN Regressor after PCA: 0.5638600866532306


In [107]:
# Linear regression after PCA
lreg = LinearRegression()
lreg.fit(X_train_reduced, y_train)
y_pred = lreg.predict(X_test_reduced)
print("R_squred score for Linear Regression after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for Linear Regression after PCA: 0.6812828438902796


In [108]:
# Ridge regression after PCA
from sklearn.linear_model import Ridge
ridge = Ridge(alpha = 10) # From project1 we know the best parameter for ridge regression is alpha = 10.
ridge.fit(X_train_reduced, y_train)
y_pred = ridge.predict(X_test_reduced)
print("R_squred score for Ridge Regression after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for Ridge Regression after PCA: 0.6798978468810819


In [109]:
# LASSO regression after PCA
from sklearn.linear_model import Lasso
lasso = Lasso(alpha = 1) # From project1 we know the best parameter for lasso regression is alpha = 1.
lasso.fit(X_train_reduced, y_train)
y_pred = lreg.predict(X_test_reduced)
print("R_squred score for Linear Regression after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for Linear Regression after PCA: 0.6812828438902796


In [110]:
# Polynomial regresssion after PCA
poly = PolynomialFeatures(2) # From project1 we know the best parameter for polynomial regression is n=2
X_train_poly_reduced = poly.fit_transform(X_train_reduced)
X_test_poly_reduced = poly.transform(X_test_reduced)
lreg = LinearRegression()
lreg.fit(X_train_poly_reduced, y_train)
y_pred = lreg.predict(X_test_poly_reduced)
print("R_squred score for Polynomial Regression after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for Polynomial Regression after PCA: -0.027996933737006735


In [111]:
# LinearSVR/SVR without kernel after PCA
from sklearn.svm import LinearSVR
lsvr = LinearSVR(C=100) # From project1 we know the best parameter for linear SVR is C=100.
lsvr.fit(X_train_reduced, y_train)
y_pred = lsvr.predict(X_test_reduced)
print("R_squred score for LinearSVR after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for LinearSVR after PCA: 0.6794504612592609


In [112]:
# SVR with rbf kernel after PCA
from sklearn.svm import SVR
svr_rbf = SVR(kernel='rbf',C=100, gamma=0.1) # From project1 we know the best parameter for SVR with rbf kernel is C=100, gamma=0.1.
svr_rbf.fit(X_train_reduced, y_train)
y_pred = svr_rbf.predict(X_test_reduced)
print("R_squred score for SVR with rbf kernel after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for SVR with rbf kernel after PCA: 0.7782835132806085


In [113]:
# SVR with poly kernel after PCA
svr_poly = SVR(kernel="poly",C=1, gamma=10) # From project1 we know the best parameter for SVR with poly kernel is C=1, gamma=10.
svr_poly.fit(X_train_reduced, y_train)
y_pred = svr_poly.predict(X_test_reduced)
print("R_squred score for SVR with poly kernel after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for SVR with poly kernel after PCA: -0.4591227911747484


In [114]:
# SVR with linear kernel after PCA
svr_poly = SVR(kernel="poly",C=100, gamma=0.001) # From project1 we know the best parameter for SVR with poly kernel is C=100, gamma=0.001.
svr_poly.fit(X_train_reduced, y_train)
y_pred = svr_poly.predict(X_test_reduced)
print("R_squred score for SVR with poly kernel after PCA: {}".format(r2_score(y_test, y_pred)))

R_squred score for SVR with poly kernel after PCA: -0.10432458430378411


## Deep Learning: Simple Neural Networks

In [115]:
X_train.shape

(468, 31)

In [116]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
#step 1: build the model
model1 = Sequential()
#input layer
model1.add(Dense(10, input_dim = 31, activation = 'sigmoid'))
#output layer: - no activation function is needed for regression
model1.add(Dense(1))
#step 2: compile the model
model1.compile(loss = 'mse', optimizer = 'sgd', metrics = ['mse'])
#step 3: train the model
model1.fit(X_train,y_train, epochs = 400, batch_size = 468)

Epoch 1/400
468/468 [==============================] - 0s 226us/sample - loss: 2407.8154 - mean_squared_error: 2407.8154
Epoch 2/400
468/468 [==============================] - 0s 4us/sample - loss: 2392.7546 - mean_squared_error: 2392.7546
Epoch 3/400
468/468 [==============================] - 0s 6us/sample - loss: 2380.0527 - mean_squared_error: 2380.0527
Epoch 4/400
468/468 [==============================] - 0s 4us/sample - loss: 2368.9661 - mean_squared_error: 2368.9661
Epoch 5/400
468/468 [==============================] - 0s 4us/sample - loss: 2359.0049 - mean_squared_error: 2359.0049
Epoch 6/400
468/468 [==============================] - 0s 4us/sample - loss: 2349.8599 - mean_squared_error: 2349.8599
Epoch 7/400
468/468 [==============================] - 0s 4us/sample - loss: 2341.3555 - mean_squared_error: 2341.3555
Epoch 8/400
468/468 [==============================] - 0s 4us/sample - loss: 2333.4033 - mean_squared_error: 2333.4033
Epoch 9/400
468/468 [=========================

Epoch 70/400
468/468 [==============================] - 0s 4us/sample - loss: 2102.1609 - mean_squared_error: 2102.1609
Epoch 71/400
468/468 [==============================] - 0s 6us/sample - loss: 2098.4751 - mean_squared_error: 2098.4751
Epoch 72/400
468/468 [==============================] - 0s 4us/sample - loss: 2094.7871 - mean_squared_error: 2094.7871
Epoch 73/400
468/468 [==============================] - 0s 6us/sample - loss: 2091.0979 - mean_squared_error: 2091.0979
Epoch 74/400
468/468 [==============================] - 0s 4us/sample - loss: 2087.4072 - mean_squared_error: 2087.4072
Epoch 75/400
468/468 [==============================] - 0s 6us/sample - loss: 2083.7156 - mean_squared_error: 2083.7156
Epoch 76/400
468/468 [==============================] - 0s 6us/sample - loss: 2080.0251 - mean_squared_error: 2080.0251
Epoch 77/400
468/468 [==============================] - 0s 6us/sample - loss: 2076.3342 - mean_squared_error: 2076.3342
Epoch 78/400
468/468 [==================

Epoch 138/400
468/468 [==============================] - 0s 4us/sample - loss: 1869.7738 - mean_squared_error: 1869.7738
Epoch 139/400
468/468 [==============================] - 0s 6us/sample - loss: 1866.8103 - mean_squared_error: 1866.8103
Epoch 140/400
468/468 [==============================] - 0s 6us/sample - loss: 1863.8595 - mean_squared_error: 1863.8595
Epoch 141/400
468/468 [==============================] - 0s 4us/sample - loss: 1860.9219 - mean_squared_error: 1860.9219
Epoch 142/400
468/468 [==============================] - 0s 9us/sample - loss: 1857.9980 - mean_squared_error: 1857.9980
Epoch 143/400
468/468 [==============================] - 0s 6us/sample - loss: 1855.0870 - mean_squared_error: 1855.0870
Epoch 144/400
468/468 [==============================] - 0s 6us/sample - loss: 1852.1891 - mean_squared_error: 1852.1891
Epoch 145/400
468/468 [==============================] - 0s 6us/sample - loss: 1849.3033 - mean_squared_error: 1849.3033
Epoch 146/400
468/468 [=========

Epoch 206/400
468/468 [==============================] - 0s 4us/sample - loss: 1694.6135 - mean_squared_error: 1694.6135
Epoch 207/400
468/468 [==============================] - 0s 6us/sample - loss: 1692.3959 - mean_squared_error: 1692.3959
Epoch 208/400
468/468 [==============================] - 0s 6us/sample - loss: 1690.1874 - mean_squared_error: 1690.1874
Epoch 209/400
468/468 [==============================] - 0s 6us/sample - loss: 1687.9880 - mean_squared_error: 1687.9880
Epoch 210/400
468/468 [==============================] - 0s 4us/sample - loss: 1685.7981 - mean_squared_error: 1685.7981
Epoch 211/400
468/468 [==============================] - 0s 4us/sample - loss: 1683.6184 - mean_squared_error: 1683.6184
Epoch 212/400
468/468 [==============================] - 0s 6us/sample - loss: 1681.4475 - mean_squared_error: 1681.4475
Epoch 213/400
468/468 [==============================] - 0s 6us/sample - loss: 1679.2859 - mean_squared_error: 1679.2859
Epoch 214/400
468/468 [=========

Epoch 274/400
468/468 [==============================] - 0s 6us/sample - loss: 1563.5581 - mean_squared_error: 1563.5581
Epoch 275/400
468/468 [==============================] - 0s 6us/sample - loss: 1561.8982 - mean_squared_error: 1561.8982
Epoch 276/400
468/468 [==============================] - 0s 6us/sample - loss: 1560.2451 - mean_squared_error: 1560.2451
Epoch 277/400
468/468 [==============================] - 0s 4us/sample - loss: 1558.5983 - mean_squared_error: 1558.5983
Epoch 278/400
468/468 [==============================] - 0s 6us/sample - loss: 1556.9589 - mean_squared_error: 1556.9589
Epoch 279/400
468/468 [==============================] - 0s 6us/sample - loss: 1555.3254 - mean_squared_error: 1555.3254
Epoch 280/400
468/468 [==============================] - 0s 6us/sample - loss: 1553.6989 - mean_squared_error: 1553.6989
Epoch 281/400
468/468 [==============================] - 0s 4us/sample - loss: 1552.0782 - mean_squared_error: 1552.0782
Epoch 282/400
468/468 [=========

Epoch 342/400
468/468 [==============================] - 0s 6us/sample - loss: 1463.6927 - mean_squared_error: 1463.6927
Epoch 343/400
468/468 [==============================] - 0s 6us/sample - loss: 1462.3823 - mean_squared_error: 1462.3823
Epoch 344/400
468/468 [==============================] - 0s 6us/sample - loss: 1461.0751 - mean_squared_error: 1461.0751
Epoch 345/400
468/468 [==============================] - 0s 6us/sample - loss: 1459.7723 - mean_squared_error: 1459.7723
Epoch 346/400
468/468 [==============================] - 0s 6us/sample - loss: 1458.4719 - mean_squared_error: 1458.4719
Epoch 347/400
468/468 [==============================] - 0s 4us/sample - loss: 1457.1749 - mean_squared_error: 1457.1749
Epoch 348/400
468/468 [==============================] - 0s 6us/sample - loss: 1455.8807 - mean_squared_error: 1455.8807
Epoch 349/400
468/468 [==============================] - 0s 4us/sample - loss: 1454.5906 - mean_squared_error: 1454.5906
Epoch 350/400
468/468 [=========

In [117]:
model1.evaluate(X_train, y_train)
model1.evaluate(X_test, y_test)

157/157 [==============================] - 0s 44us/sample - loss: 43.5710 - mean_squared_error: 43.5710


[43.57096940089183, 43.57097]

In [118]:
y_pred_train = model1.predict(X_train)
y_pred_test = model1.predict(X_test)
print('R_squred training score for neutral network: {}'.format(r2_score(y_train, y_pred_train)))
print('R_squred testing score for neutral network: {}'.format(r2_score(y_test, y_pred_test)))

R_squred training score for neutral network: 0.3983534887482504
R_squred testing score for neutral network: 0.9017681120520391


## Deep Learning: Neural Networks with Multiple Layers

In [119]:
#step 1: build model
model2 = Sequential()
#input layer
model2.add(Dense(20, input_dim = 31, activation = 'sigmoid'))
#hidden layers
model2.add(Dense(10, activation = 'relu'))
model2.add(Dense(5, activation = 'relu'))
#output layer
model2.add(Dense(1))
#step 2: compile the model
model2.compile(loss= 'mse' , optimizer = 'sgd',metrics = ['mse'] )
#step 3: train the model
model2.fit(X_train, y_train, epochs = 400, batch_size = 468)

Epoch 1/400
468/468 [==============================] - 0s 299us/sample - loss: 2400.6396 - mean_squared_error: 2400.6396
Epoch 2/400
468/468 [==============================] - 0s 6us/sample - loss: 2389.9214 - mean_squared_error: 2389.9214
Epoch 3/400
468/468 [==============================] - 0s 6us/sample - loss: 2385.0105 - mean_squared_error: 2385.0105
Epoch 4/400
468/468 [==============================] - 0s 4us/sample - loss: 2379.0669 - mean_squared_error: 2379.0669
Epoch 5/400
468/468 [==============================] - 0s 4us/sample - loss: 2370.6489 - mean_squared_error: 2370.6489
Epoch 6/400
468/468 [==============================] - 0s 6us/sample - loss: 2357.0120 - mean_squared_error: 2357.0120
Epoch 7/400
468/468 [==============================] - 0s 4us/sample - loss: 2333.9690 - mean_squared_error: 2333.9690
Epoch 8/400
468/468 [==============================] - 0s 6us/sample - loss: 2305.6394 - mean_squared_error: 2305.6394
Epoch 9/400
468/468 [=========================

Epoch 70/400
468/468 [==============================] - 0s 6us/sample - loss: 2319.6399 - mean_squared_error: 2319.6399
Epoch 71/400
468/468 [==============================] - 0s 6us/sample - loss: 2319.2207 - mean_squared_error: 2319.2207
Epoch 72/400
468/468 [==============================] - 0s 6us/sample - loss: 2318.8179 - mean_squared_error: 2318.8179
Epoch 73/400
468/468 [==============================] - 0s 6us/sample - loss: 2318.4319 - mean_squared_error: 2318.4319
Epoch 74/400
468/468 [==============================] - 0s 4us/sample - loss: 2318.0605 - mean_squared_error: 2318.0605
Epoch 75/400
468/468 [==============================] - 0s 6us/sample - loss: 2317.7041 - mean_squared_error: 2317.7041
Epoch 76/400
468/468 [==============================] - 0s 6us/sample - loss: 2317.3625 - mean_squared_error: 2317.3625
Epoch 77/400
468/468 [==============================] - 0s 4us/sample - loss: 2317.0332 - mean_squared_error: 2317.0332
Epoch 78/400
468/468 [==================

Epoch 138/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.7380 - mean_squared_error: 2309.7380
Epoch 139/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.7112 - mean_squared_error: 2309.7112
Epoch 140/400
468/468 [==============================] - 0s 5us/sample - loss: 2309.6858 - mean_squared_error: 2309.6858
Epoch 141/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.6609 - mean_squared_error: 2309.6609
Epoch 142/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.6367 - mean_squared_error: 2309.6367
Epoch 143/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.6140 - mean_squared_error: 2309.6140
Epoch 144/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.5925 - mean_squared_error: 2309.5925
Epoch 145/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.5715 - mean_squared_error: 2309.5715
Epoch 146/400
468/468 [=========

Epoch 206/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.1035 - mean_squared_error: 2309.1035
Epoch 207/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.1021 - mean_squared_error: 2309.1021
Epoch 208/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.1003 - mean_squared_error: 2309.1003
Epoch 209/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0989 - mean_squared_error: 2309.0989
Epoch 210/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.0972 - mean_squared_error: 2309.0972
Epoch 211/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0959 - mean_squared_error: 2309.0959
Epoch 212/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0940 - mean_squared_error: 2309.0940
Epoch 213/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0925 - mean_squared_error: 2309.0925
Epoch 214/400
468/468 [=========

Epoch 274/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0632 - mean_squared_error: 2309.0632
Epoch 275/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0630 - mean_squared_error: 2309.0630
Epoch 276/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0625 - mean_squared_error: 2309.0625
Epoch 277/400
468/468 [==============================] - 0s 9us/sample - loss: 2309.0632 - mean_squared_error: 2309.0632
Epoch 278/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0627 - mean_squared_error: 2309.0627
Epoch 279/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0625 - mean_squared_error: 2309.0625
Epoch 280/400
468/468 [==============================] - 0s 9us/sample - loss: 2309.0623 - mean_squared_error: 2309.0623
Epoch 281/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0625 - mean_squared_error: 2309.0625
Epoch 282/400
468/468 [=========

Epoch 342/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0603 - mean_squared_error: 2309.0603
Epoch 343/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0598 - mean_squared_error: 2309.0598
Epoch 344/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.0601 - mean_squared_error: 2309.0601
Epoch 345/400
468/468 [==============================] - 0s 4us/sample - loss: 2309.0605 - mean_squared_error: 2309.0605
Epoch 346/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0601 - mean_squared_error: 2309.0601
Epoch 347/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0603 - mean_squared_error: 2309.0603
Epoch 348/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0601 - mean_squared_error: 2309.0601
Epoch 349/400
468/468 [==============================] - 0s 6us/sample - loss: 2309.0598 - mean_squared_error: 2309.0598
Epoch 350/400
468/468 [=========

In [120]:
model2.evaluate(X_train, y_train)
model2.evaluate(X_test, y_test)

157/157 [==============================] - 0s 38us/sample - loss: 447.4989 - mean_squared_error: 447.4988


[447.49887235605036, 447.49884]

In [121]:
y_pred_train = model2.predict(X_train)
y_pred_test = model2.predict(X_test)
print('R_squred training score for multiple layer neutral network: {}'.format(r2_score(y_train, y_pred_train)))
print('R_squred testing score for multiple layer neutral network: {}'.format(r2_score(y_test, y_pred_test)))

R_squred training score for multiple layer neutral network: -7.115872335816675e-09
R_squred testing score for multiple layer neutral network: -0.008897861863102152


# Classification

In [122]:
X = df.drop(['Audit_Risk', 'Risk'], axis = 1)
y = df['Risk']
from sklearn.model_selection import train_test_split
X_train_org, X_test_org, y_train, y_test = train_test_split(X,y, random_state = 0)
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train_org)
X_test = scaler.transform(X_test_org)

## Hard voting and soft voting from KNN classifier, logit regression and svc with rbf kernel.

In [123]:
### Hard voting 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
log_clf = LogisticRegression()
log_clf.fit(X_train, y_train)
knn_clf = KNeighborsClassifier(3)
knn_clf.fit(X_train, y_train)
svm_clf = SVC(kernel='rbf', C=10, gamma=1, probability = True)
svm_clf.fit(X_train, y_train)

Hvoting_clf = VotingClassifier(estimators=[('lr', log_clf), ('knn', knn_clf), ('svc', svm_clf)], voting='hard')
Hvoting_clf.fit(X_train, y_train)

from sklearn.metrics import accuracy_score
for clf in (log_clf, knn_clf, svm_clf, Hvoting_clf):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, accuracy_score(y_test, y_pred))


LogisticRegression 0.9936305732484076
KNeighborsClassifier 1.0
SVC 1.0
VotingClassifier 1.0


C:\Users\nicol\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


In [124]:
### Soft voting 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
log_clf = LogisticRegression()
log_clf.fit(X_train, y_train)
knn_clf = KNeighborsClassifier(3)
knn_clf.fit(X_train, y_train)
svm_clf = SVC(kernel='rbf', C=10, gamma=1, probability = True)
svm_clf.fit(X_train, y_train)

Svoting_clf = VotingClassifier(estimators=[('lr', log_clf), ('knn', knn_clf), ('svc', svm_clf)], voting='soft')
Svoting_clf.fit(X_train, y_train)

from sklearn.metrics import accuracy_score
for clf in (log_clf, knn_clf, svm_clf, Svoting_clf):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, accuracy_score(y_test, y_pred))


LogisticRegression 0.9936305732484076
KNeighborsClassifier 1.0
SVC 1.0
VotingClassifier 1.0


C:\Users\nicol\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


## Bagging and Pasting on KNN Classifier and Decision Tree

In [125]:
# Bagging on KNN Classifier
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

param = {'n_estimators': [100, 200, 300, 400, 500]}

knn_clf = KNeighborsClassifier(3) # From project1 we know that the best parameter for KNN classifier is n=3.
bag_knn = BaggingClassifier(knn_clf, bootstrap=True, random_state=0)
grid_search = GridSearchCV(bag_knn, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))
    

Best parameters: {'n_estimators': 100}
Best cross-validation score: 1.00


In [126]:
bag_knn1 = BaggingClassifier(knn_clf, n_estimators=100, bootstrap=True, random_state=0)
bag_knn1.fit(X_train, y_train)
y_test_pred = bag_knn1.predict(X_test)
from sklearn.metrics import accuracy_score
print("Accuracy score for bagging knn classifier with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for bagging knn classifier with best parameters: 1.0


In [127]:
# Bagging on Decision Tree
from sklearn.tree import DecisionTreeClassifier
param = {'n_estimators': [100, 200, 300, 400, 500]}
dtree_clf = DecisionTreeClassifier()
bag_dtree = BaggingClassifier(dtree_clf, bootstrap=True, random_state=0)
grid_search = GridSearchCV(bag_dtree, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'n_estimators': 100}
Best cross-validation score: 1.00


In [128]:
bag_dtree1 = BaggingClassifier(dtree_clf, n_estimators=100, bootstrap=True, random_state=0)
bag_dtree1.fit(X_train, y_train)
y_test_pred = bag_dtree1.predict(X_test)
print("Accuracy score for bagging decision tree with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for bagging decision tree with best parameters: 1.0


In [129]:
# Pasting on KNN Classifier
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

param = {'n_estimators': [100, 200, 300, 400, 500]}

knn_clf = KNeighborsClassifier(3) # From project1 we know that the best parameter for KNN classifier is n=3.
bag_knn = BaggingClassifier(knn_clf, bootstrap=False, random_state=0)
grid_search = GridSearchCV(bag_knn, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))
  

Best parameters: {'n_estimators': 100}
Best cross-validation score: 1.00


In [130]:
past_knn1 = BaggingClassifier(knn_clf, n_estimators=100, bootstrap=False, random_state=0)
past_knn1.fit(X_train, y_train)
y_test_pred = past_knn1.predict(X_test)
from sklearn.metrics import precision_score
print("Accuracy score for pasting knn classifier with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for pasting knn classifier with best parameters: 1.0


In [131]:
# Pasting on Decision Tree
from sklearn.tree import DecisionTreeClassifier
param = {'n_estimators': [100, 200, 300, 400, 500]}
dtree_clf = DecisionTreeClassifier()
bag_dtree = BaggingClassifier(dtree_clf, bootstrap=False, random_state=0)
grid_search = GridSearchCV(bag_dtree, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'n_estimators': 100}
Best cross-validation score: 1.00


In [132]:
bag_dtree1 = BaggingClassifier(dtree_clf, n_estimators=100, bootstrap=False, random_state=0)
bag_dtree1.fit(X_train, y_train)
y_test_pred = bag_dtree1.predict(X_test)
from sklearn.metrics import precision_score
print("Accuracy score for pasting decision tree with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for pasting decision tree with best parameters: 1.0


## Adaboost boosting on Decision Tree and Logistic Regression

In [133]:
# Adaboost on Decision Tree Classifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
dtree_clf = DecisionTreeClassifier()
param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
ada_dtree = AdaBoostClassifier(dtree_clf, random_state=0)
grid_search = GridSearchCV(ada_dtree, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.2, 'n_estimators': 100}
Best cross-validation score: 1.00


In [134]:
ada_dtree1 = AdaBoostClassifier(dtree_clf, n_estimators=100, learning_rate=0.2, random_state=0)
ada_dtree1.fit(X_train, y_train)
y_test_pred = ada_dtree1.predict(X_test)
print("Accuracy score for adaboost decision tree classifier with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for adaboost decision tree classifier with best parameters: 1.0


In [135]:
# Adaboost on Logit Regression
from sklearn.linear_model import LogisticRegression
lg_clf = LogisticRegression(penalty = 'l2', C = 0.1) # From project1 we know that the best pramaters for logit regression is C=0.1 and penalty=l2.
param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
ada_lg = AdaBoostClassifier(lg_clf, random_state=0)
grid_search = GridSearchCV(ada_lg, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.2, 'n_estimators': 100}
Best cross-validation score: 1.00


In [136]:
ada_lg1 = AdaBoostClassifier(lg_clf, n_estimators=100, learning_rate=0.2, random_state=0)
ada_lg1.fit(X_train, y_train)
y_test_pred = ada_lg1.predict(X_test)
print("Accuracy score for adaboost logit regression with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for adaboost logit regression with best parameters: 1.0


## Gradient Boosting Model(Decision Tree Classifier)

In [137]:
from sklearn.ensemble import GradientBoostingClassifier
param = {"n_estimators": [100, 200, 300, 400, 500],
        "learning_rate":[0.2, 0.4, 0.6, 0.8, 1]}
gbrt = GradientBoostingClassifier(random_state=0)
grid_search = GridSearchCV(gbrt, param, cv=5, return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best parameters: {}".format(grid_search.best_params_))
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))

Best parameters: {'learning_rate': 0.2, 'n_estimators': 100}
Best cross-validation score: 1.00


In [138]:
gbrt_clf1 = GradientBoostingClassifier(n_estimators=100, learning_rate=0.2, random_state=0)
gbrt_clf1.fit(X_train, y_train)
y_test_pred = gbrt_clf1.predict(X_test)
print("Accuracy score for gradient boosting classifier with best parameters: {}".format(accuracy_score(y_test, y_test_pred)))

Accuracy score for gradient boosting classifier with best parameters: 1.0


## PCA and Classification with best parameters

#### We are using the X_train_reduced and X_test_reduced from the regression since both classification and regression share the same feature data set. And we already found the best number of dimension with grid search from previous work.


In [139]:
# KNN classifier after PCA
from sklearn.neighbors import KNeighborsClassifier
knn_clf3 = KNeighborsClassifier(3) # From project1 we know the best parameter for knn classifier is k=3.
knn_clf3.fit(X_train_reduced, y_train)
y_pred = knn_clf3.predict(X_test_reduced)
print("Accuracy score for KNN Classifier after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for KNN Classifier after PCA: 1.0


In [140]:
# Logic regressoin after PCA
lg_clf1 = LogisticRegression(C=0.1, penalty='l2') # From project1 we know the best parameter for logit regression is C=0.1, penalty=l2.
lg_clf1.fit(X_train_reduced, y_train)
y_pred = lg_clf1.predict(X_test_reduced)
print("Accuracy score for logit regression after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for logit regression after PCA: 0.9554140127388535


In [141]:
# LinearSVC after PCA
from sklearn.svm import LinearSVC
SVC1 = LinearSVC(C=1) # From project1 we know the best parameter for LinearSVC is C=1.
SVC1.fit(X_train_reduced, y_train)
y_pred = SVC1.predict(X_test_reduced)
print("Accuracy score for LinearSVC after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for LinearSVC after PCA: 1.0


In [142]:
# SVC with rbf kernel after PCA
from sklearn.svm import SVC
SVC_rbf1 = SVC(kernel='rbf', C=10, gamma=1) # From project1 we know the best parameter for SVC with rbf kernel is C=10, gamma=1.
SVC_rbf1.fit(X_train_reduced, y_train)
y_pred = SVC_rbf1.predict(X_test_reduced)
print("Accuracy score for SVC with rbf kernel after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for SVC with rbf kernel after PCA: 1.0


In [143]:
# SVC with poly kernel after PCA
from sklearn.svm import SVC
SVC_poly1 = SVC(kernel='poly', C=100, gamma=0.1) # From project1 we know the best parameter for SVC with rbf kernel is C=100, gamma=0.1.
SVC_poly1.fit(X_train_reduced, y_train)
y_pred = SVC_poly1.predict(X_test_reduced)
print("Accuracy score for SVC with poly kernel after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for SVC with poly kernel after PCA: 0.9617834394904459


In [144]:
# SVC with linear kernel after PCA
from sklearn.svm import SVC
SVC_ln1 = SVC(kernel='linear', C=0.1, gamma=0.001) # From project1 we know the best parameter for SVC with linear kernel is C=0.1, gamma=0.001.
SVC_ln1.fit(X_train_reduced, y_train)
y_pred = SVC_ln1.predict(X_test_reduced)
print("Accuracy score for SVC with linear kernel after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for SVC with linear kernel after PCA: 0.9872611464968153


In [145]:
# Decision tree after PCA
from sklearn.tree import DecisionTreeClassifier
dtree1 = DecisionTreeClassifier() # From project1 we know the best parameter for SVC with linear kernel is C=0.1, gamma=0.001.
dtree1.fit(X_train_reduced, y_train)
y_pred = dtree1.predict(X_test_reduced)
print("Accuracy score for decision tree after PCA: {}".format(accuracy_score(y_test, y_pred)))

Accuracy score for decision tree after PCA: 1.0


## Deep Learning: Simple Neural Networks

In [146]:
X_train.shape

(468, 31)

In [147]:
#step 1: build model
model1 = Sequential()
#input layer
model1.add(Dense(10, input_dim = 31, activation = 'relu'))
#output layer
model1.add(Dense(1, activation = 'sigmoid'))
#step 2: make computational graph - compile
model1.compile(loss= 'binary_crossentropy' , optimizer = 'adam',metrics = ['accuracy'] )
#step 3: train the model - fit
model1.fit(X_train, y_train, epochs = 40, batch_size = 468)

Epoch 1/40
468/468 [==============================] - 0s 475us/sample - loss: 0.6896 - acc: 0.5620
Epoch 2/40
468/468 [==============================] - 0s 4us/sample - loss: 0.6866 - acc: 0.5876
Epoch 3/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6835 - acc: 0.6047
Epoch 4/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6805 - acc: 0.6090
Epoch 5/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6776 - acc: 0.6282
Epoch 6/40
468/468 [==============================] - 0s 4us/sample - loss: 0.6746 - acc: 0.6410
Epoch 7/40
468/468 [==============================] - 0s 4us/sample - loss: 0.6716 - acc: 0.6581
Epoch 8/40
468/468 [==============================] - 0s 4us/sample - loss: 0.6686 - acc: 0.6624
Epoch 9/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6657 - acc: 0.6624
Epoch 10/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6628 - acc: 0.6859
Epoch 11/40
468/468 [======

In [148]:
model1.evaluate(X_train, y_train)
model1.evaluate(X_test, y_test)

157/157 [==============================] - 0s 38us/sample - loss: 0.5755 - acc: 0.9554


[0.5754925566873733, 0.955414]

In [149]:
y_pred_train = model1.predict(X_train)
y_pred_test = model1.predict(X_test)
y_train_pred = np.where(y_pred_train >= 0.5, 1, 0)
y_test_pred = np.where(y_pred_test >= 0.5, 1, 0)
print('Training accuracy score for simple neutral network: {}'.format(accuracy_score(y_train, y_train_pred)))
print('Testing accuracy score for simple neutral network: {}'.format(accuracy_score(y_test, y_test_pred)))

Training accuracy score for simple neutral network: 0.9722222222222222
Testing accuracy score for simple neutral network: 0.9554140127388535


## Deep Learning: Neural Networks with Multiple Layers

In [150]:
#step 1: build model
model2 = Sequential()
#input layer
model2.add(Dense(20, input_dim = 31, activation = 'relu'))
#hidden layers
model2.add(Dense(10, activation = 'relu'))
model2.add(Dense(5, activation = 'relu'))
#output layer
model2.add(Dense(1, activation = 'sigmoid'))
#step 2: compile the model
model2.compile(loss= 'binary_crossentropy' , optimizer = 'adam',metrics = ['accuracy'] )
#step 3: train the model
model2.fit(X_train, y_train, epochs = 40, batch_size = 468)

Epoch 1/40
468/468 [==============================] - 0s 714us/sample - loss: 0.6878 - acc: 0.6261
Epoch 2/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6864 - acc: 0.6474
Epoch 3/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6848 - acc: 0.6688
Epoch 4/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6832 - acc: 0.6752
Epoch 5/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6815 - acc: 0.6923
Epoch 6/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6795 - acc: 0.7265
Epoch 7/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6774 - acc: 0.7479
Epoch 8/40
468/468 [==============================] - 0s 4us/sample - loss: 0.6753 - acc: 0.7585
Epoch 9/40
468/468 [==============================] - 0s 6us/sample - loss: 0.6730 - acc: 0.7863
Epoch 10/40
468/468 [==============================] - 0s 9us/sample - loss: 0.6705 - acc: 0.7842
Epoch 11/40
468/468 [======

In [151]:
model2.evaluate(X_train, y_train)
model2.evaluate(X_test, y_test)

157/157 [==============================] - 0s 38us/sample - loss: 0.5520 - acc: 0.9172


[0.5519759844822488, 0.91719747]

In [152]:
y_pred_train = model2.predict(X_train)
y_pred_test = model2.predict(X_test)
y_train_pred = np.where(y_pred_train >= 0.5, 1, 0)
y_test_pred = np.where(y_pred_test >= 0.5, 1, 0)
print('Training accuracy score for neutral network with multiple layers: {}'.format(accuracy_score(y_train, y_train_pred)))
print('Testing accuracy score for neutral network with multiple layers: {}'.format(accuracy_score(y_test, y_test_pred)))

Training accuracy score for neutral network with multiple layers: 0.905982905982906
Testing accuracy score for neutral network with multiple layers: 0.9171974522292994


# Final Report

In [153]:
Regression = pd.read_csv('Regression_Results.csv')
Regression

Model Name     Best Parameters  \
0                                      Knn Regression       n_neighbors=3   
1                                   Linear Regression                 NaN   
2                                    Ridge Regression            alpha=10   
3                                    Lasso Regression             alpha=1   
4                               Polynomial Regression            degree=2   
5                                           LinearSVR               C=100   
6                                   SVR(kenerl='rbf')    C=100, gamma=0.1   
7                                  SVR(kenerl='poly')       C=1, gamma=10   
8                                SVR(kenerl='linear')  C=100, gamma=0.001   
9   Gridient Boosting Regressor(on Decision Tree R...                 NaN   
10                              Simple Neural Network                 NaN   
11                                  ML Neural Network                 NaN   

    Training Score  Testing Score  Testing R_Square  \
0            0.352          0.444             0.626   
1           -1.593          0.697             0.444   
2            0.404         -0.995             0.662   
3            0.525         -0.563             0.697   
4            0.911          0.963             0.999   
5            0.644          0.452             0.687   
6            0.696          0.369             0.797   
7            0.945          0.829             0.975   
8            0.646          0.436             0.689   
9              NaN            NaN             0.898   
10             NaN            NaN             0.872   
11             NaN            NaN            -0.009   

    Testing R_Squre after PCA  Testing R_Squre after Bagging  \
0                       0.564                            NaN   
1                       0.681                          0.604   
2                       0.680                            NaN   
3                       0.681                            NaN   
4                      -0.028                          0.998   
5                       0.679                            NaN   
6                       0.778                            NaN   
7                      -0.459                            NaN   
8                      -0.104                            NaN   
9                         NaN                            NaN   
10                        NaN                            NaN   
11                        NaN                            NaN   

    Testing R_Squre after Pasting  Testing R_Square after Adaboost boosting  
0                             NaN                                       NaN  
1                           0.444                                     0.318  
2                             NaN                                       NaN  
3                             NaN                                       NaN  
4                           0.999                                     0.999  
5                             NaN                                       NaN  
6                             NaN                                       NaN  
7                             NaN                                       NaN  
8                             NaN                                       NaN  
9                             NaN                                       NaN  
10                            NaN                                       NaN  
11                            NaN                                       NaN

From this table, we can see that most models have a lower R_squared testing score after dimension reduction with PCA, indicating a better model performance without PCA. Only linear regression and ridge regression have improved testing R_squred scores. The linear regression's testing R_square improved from 0.444 to 0.681 after PCA. The ridge regression's testing R_square improved from 0.662 to 0.680.

In terms of ensembling, only bagging on linear regressoin improved the model performace(0.604 to 0.444). None of the ensembling methods improved the performance of polynomial regression.

For the deep learning models, the netural network without hidden layer has a better performance with a testing R_square of 0.872.

Among all of these models, polynomial regression with two degree without PCA has the best performace, with a testing R_squared score of 0.999.

In [154]:
Classification = pd.read_csv('Classification_Results.csv')
Classification

Model Name     Best Parameters  Training Score  Testing Score  \
0          Knn Classifier       n_neighbors=3           0.998          1.000   
1        Logic Regression   C=0.1, penalty=l2           0.987          0.981   
2               LinearSVC                 C=1           1.000          1.000   
3       SVC(kenerl='rbf')       C=10, gamma=1           0.998          1.000   
4      SVC(kenerl='poly')    C=100, gamma=0.1           0.989          0.968   
5    SVC(kenerl='linear')  C=0.1, gamma=0.001           0.991          0.981   
6           Decision Tree                 NaN           1.000          1.000   
7  Simple Neural Networks                 NaN           0.912            NaN   
8      ML Neural Networks                 NaN           1.000            NaN   

   Testing Accuracy  Testing Accuracy after PCA  \
0             1.000                       1.000   
1             0.981                       0.955   
2             1.000                       1.000   
3             1.000                       1.000   
4             0.981                       0.963   
5             0.987                       0.987   
6             1.000                       1.000   
7               NaN                         NaN   
8               NaN                         NaN   

   Testing Accuracy after Bagging  Testing Accuracy after Pasting  \
0                             1.0                             1.0   
1                             NaN                             NaN   
2                             NaN                             NaN   
3                             NaN                             NaN   
4                             NaN                             NaN   
5                             NaN                             NaN   
6                             1.0                             1.0   
7                             NaN                             NaN   
8                             NaN                             NaN   

   Testing Accuracy after Adaboost boosting  Testing Accuracy after  GDB  
0                                       NaN                          NaN  
1                                       1.0                          NaN  
2                                       NaN                          NaN  
3                                       NaN                          NaN  
4                                       NaN                          NaN  
5                                       NaN                          NaN  
6                                       1.0                          1.0  
7                                       NaN                          NaN  
8                                       NaN                          NaN

From this table, we can see that none of the classification models got improved accuracy score after PCA, indicating a better model performance without PCA. There are 4 models have a 100% testing accuracy with and without PCA, including KNeighbor Classifier, LinearSVC, SVC with rbf kernel and decision tree. 

In terms of ensembling, none of the applied ensembling methods made any difference on the model performance.

For the deep learning models, the netural network without hidden layer has a testing accuracy of 0.912, less that the netural network with multiple layers with a testing accuracy of 1. In this case, the ML neural network performed better. 
